[GlobalISel] Lowering of {get,set,reset}_fpenv

spavloff opened the pull request:
The intrinsics get_fpenv, set_fpenv and reset_fpenv in this change are implemented as calls to math library functions. Target specific lowering will be implemented later on.

spavloff committed:
[GlobalISel] Lowering of {get,set,reset}_fpenv

The intrinsics get_fpenv, set_fpenv and reset_fpenv in this change are
implemented as calls to math library functions. Target specific lowering
will be implemented later on.

tschuett commented on llvm/lib/Target/AArch64/GISel/AArch64LegalizerInfo.cpp:
```diff
@@ -1122,6 +1122,8 @@ AArch64LegalizerInfo::AArch64LegalizerInfo(const AArch64Subtarget &ST)
   getActionDefinitionsBuilder(G_FMAD).lower();
 
   // Access to floating-point environment.
+  getActionDefinitionsBuilder({G_GET_FPENV, G_SET_FPENV, G_RESET_FPENV})
```
You could merge them into one builder. They have similar concepts and similar legalization strategies.

spavloff committed:
Fix code formatting

- fix clang-format errors,
- merge builders for FP environment.

tschuett commented on llvm/test/CodeGen/AArch64/fpenv.ll:
```diff
@@ -62,4 +62,47 @@ define void @func_set_rounding_downward() {
   ret void
 }
 
+define i64 @get_fpenv_01() nounwind {
```
This is presumably not testing what you planned. The RUN line does not contain the string `globalisel`. It is testing DAG code generation. If you look at the sister file xtn.ll, it has two RUN lines to test code generation for both selectors and highlight their differences.

spavloff committed:
Fix tests

spavloff committed:
Fix tests again

- Tests for *_fpenv intrinsics are moved to a separate file in the
  directory AArch64/GlobalISel.
- Update the test TableGen/GlobalISelEmitter.td

aemerson reviewed (approved):
Looks ok, if there's an existing codegen test in AArch64 for these intrinsic could you add RUN lines for GISel?

arsenm commented on llvm/test/CodeGen/AArch64/GlobalISel/legalize-fpenv.mir:
```diff
@@ -0,0 +1,88 @@
+  declare void @llvm.set.fpenv.i64(i64)
+  declare void @llvm.reset.fpenv()
+
+  define i64 @func_get_fpenv() {
```
Don't need the IR section

arsenm commented on llvm/test/CodeGen/AArch64/GlobalISel/legalize-fpenv.mir:
```diff
@@ -0,0 +1,88 @@
+name:            func_get_fpenv
+tracksRegLiveness: true
+body:             |
+  bb.1.entry:
```
bb.0, drop .1.entry

spavloff committed:
Fix legalization test

spavloff committed:
Fix tests after rebase

spavloff committed:
Update test

Call to fesetenv is now tail-optimized.
